src/Reopt/TypeInference/Solver/Solver.hs

```diff
@@ -82,74 +81,99 @@ traceContext description action = do
-processAtomicConstraints = traceContext "processAtomicConstraints" $ do
-  dequeueEqC >>= \case
-    Just c  -> solveEqC c >> processAtomicConstraints
-    Nothing -> dequeueEqRowC >>= \case
-      Just c  -> solveEqRowC c >> processAtomicConstraints
-      Nothing -> condEqSolver >>= \case
-        True -> processAtomicConstraints
-        False -> pure ()
-  where
-    -- This solver will solve one at a time, which is important to get
-    -- around the +p++ case.  We solve a single ptr add, then propagate
-    -- the new eq and roweq constraints.
-    condEqSolver = do
-      ceqs <- field @"ctxCondEqs" <<.= mempty -- get constraints and c
-      go [] ceqs
+traceContext' :: PP.Pretty v => PP.Doc () -> v -> SolverM a -> SolverM a
+traceContext' msg v = traceContext (msg <> ": " <> PP.pretty v)
 
-    restore :: [Conditional'] -> SolverM ()
-    restore cs = field @"ctxCondEqs" .= cs
+--------------------------------------------------------------------------------
+-- Solver loop
+
+data Retain   = Retain | Discard
+  deriving Eq
+
+data Progress = Progress | NoProgress
+  deriving Eq
+
+madeProgress :: Progress -> Bool
+madeProgress Progress = True
+madeProgress _        = False
+
+solveHead :: Lens' ConstraintSolvingState [a] ->
+             (a -> SolverM ()) ->
+             SolverM Progress
+solveHead fld doit = do
+  v <- popField fld
+  case v of
+    Nothing -> pure NoProgress
+    Just v' -> doit v' $> Progress
+
+solveFirst :: Lens' ConstraintSolvingState [a] ->
+              (a -> SolverM (Retain, Progress)) ->
+              SolverM Progress
+solveFirst fld solve = do
+  cstrs <- fld <<.= [] -- get constraints and c
+  go [] cstrs
+  where
+    restore cs = fld %= (++ cs)
 
-    -- FIXME: we might want to drop conditionals when they are never
+    -- FIXME: we might want to drop constraints when they are never
     -- going to be satisfiable.
-    go acc [] = restore acc $> False -- finished here, we didn't so anything.
+    go acc [] = restore acc $> NoProgress -- finished here, we didn't so anything.
     go acc (c : cs) = do
-      solved <- solveConditional c
-      if solved
-        -- Conditional fired, remove it and continue solving
-        then restore (cs ++ acc) $> True
-        -- Conditional couldn't be fired, try next conditionals
-        else go (c : acc) cs
+      (retain, progress) <- solve c
+      let acc' = if retain == Retain then c : acc else acc
+      if madeProgress progress
+        then restore (cs ++ acc') $> Progress
+        else go acc' cs
+
+solverLoop :: SolverM ()
+solverLoop = do
+  keepGoing <- foldr once (pure False) solvers
+  when keepGoing solverLoop
+  where
+    solvers = [ solveHead  (field @"ctxEqCs")    solveEqC
+              , solveHead  (field @"ctxEqRowCs") solveEqRowC
+              , solveFirst (field @"ctxCondEqs") solveConditional
+              ]
+
+    once m rest = do
+      progress <- m
+      if madeProgress progress then pure True else rest
```

Comment:
You could use `Control.Monad.Extra.ifM`.

Reply:
We already depend on extra, so we could even change the types and use ```||^``` which is really what is going on here